templates/android/library/src/main/java/io/appwrite/services/Realtime.kt.twig

```diff
@@ -0,0 +1,165 @@
+    }
+
+    fun subscribe(
+        vararg channels: String,
+        callback: (Any) -> Unit
+    ) : RealtimeSubscription {
+        channels.forEach {
+            if (!channelCallbacks.containsKey(it)) {
+                channelCallbacks[it] = mutableListOf(callback)
+                return@forEach
+            }
+            channelCallbacks[it]?.add(callback)
+        }
+
+        launch {
+            subCallDepth++
+            delay(DEBOUNCE_MILLIS)
+            if (subCallDepth == 1) {
+                createSocket()
+            }
+            subCallDepth--
+        }
+
+        return RealtimeSubscription { unsubscribe() }
+    }
+
+    fun unsubscribe() {
+        channelCallbacks = mutableMapOf()
+        errorCallbacks = mutableSetOf()
+        closeSocket()
+    }
+
+    fun doOnError(callback: (RealtimeError) -> Unit) {
+        errorCallbacks.add(callback)
+    }
+
+    private inner class {{ spec.title | caseUcfirst }}WebSocketListener : WebSocketListener() {
+
+        override fun onOpen(webSocket: WebSocket, response: Response) {
+            super.onOpen(webSocket, response)
+            print("WebSocket connected.")
+        }
+
+        override fun onMessage(webSocket: WebSocket, text: String) {
+            super.onMessage(webSocket, text)
+            print("WebSocket message received: $text")
+
+            launch(IO) {
+                val message = parse<RealtimeMessage>(text)
+                if (message?.channels == null) {
+                    val error = parse<RealtimeError>(text) ?: return@launch
+                    errorCallbacks.forEach {
+                        it.invoke(error)
+                    }
+                    return@launch
+                }
+
+                message.channels.forEachAsync { channel ->
+                    channelCallbacks[channel]?.forEachAsync { callback ->
+                        callback.invoke(message.payload)
+                    }
+                }
+            }
+        }
+
+        override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
+            super.onClosing(webSocket, code, reason)
+            print("WebSocket closing with code $code because: $reason.")
+            if (code == RealtimeCodes.POLICY_VIOLATION.value) {
+                return
+            }
+            print("Reconnecting in 1 second.")
+            launch {
+                delay(1000)
+                createSocket()
+            }
+        }
+
+        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
+            super.onFailure(webSocket, t, response)
+            print("WebSocket failure.")
+            t.printStackTrace()
+        }
+
+        private inline fun <reified T> parse(text: String): T? {
+            return try {
+                Gson().fromJson(text, T::class.java)
+            } catch (ex: JsonSyntaxException) {
+                ex.printStackTrace()
+                null
+            } catch (ex: JsonParseException) {
```

Comment:
Are these the only two possible exceptions that we want to catch ? Shouldn't we just catch the general `Exception` and print stack trace and return `null` ?